File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <set>
#include <iostream>

using namespace std;

int answer;
bool visited[7];
set<int> results;
bool check_decimal(int i);
void dfs(string numbers, string current);
int solution(string numbers) {
    dfs(numbers, "");
    for(int i : results) if(check_decimal(i))answer++;
    return answer;
}

void dfs(string numbers, string current){
    if(!current.empty()) results.insert(stoi(current));
    
    for(int i=0; i<numbers.length(); i++){
        if(visited[i]) continue;
        
        visited[i] = true;
        dfs(numbers, current+numbers[i]);
        visited[i] = false;
    }
}
bool check_decimal(int i){
    if(i==0) return false;
    if(i==1) return false;
    if(i==2) return true;
    else {
        for(int k=2; k<i; k++) {
            if(i%k==0) return false;
        };
        return true;
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    public static Set<Integer> set = new HashSet<>();
    
    public int solution(String numbers) {
        char[] arr = numbers.toCharArray();
        boolean[] visited = new boolean[arr.length];
        int answer = 0;
        
        dfs("", arr, visited);
        
        for(int i : set){
            if(isPrime(i)) answer++;
        }
        
        System.out.println(answer);
        
        return answer;
    }
    public static void dfs(String current, char[] arr, boolean[] visited){
        if(!current.isEmpty()) {
            set.add(Integer.parseInt(current));
        }
        for(int i=0; i<arr.length; i++){
            if(!visited[i]){
                visited[i] = true;
                dfs(current+arr[i], arr, visited);
                visited[i] = false;
            }
        }
    }
    
    public static boolean isPrime(int num){
        if(num == 0) return false;
        if(num == 1) return false;
        if(num == 2) return true;
        
        for(int i=2; i<num; i++){
            if(num%i==0) return false;
        }
        return true;
    }
}